dcicutils/project_utils.py

```diff
@@ -0,0 +1,453 @@
+        app_project: Project = cls.app_project  # Now that it's initialized, make sure it comes from the right place
+        return app_project
+
+    @classmethod
+    def show_herald(cls):
+        app_project = cls.PROJECT_BASE_CLASS.app_project_maker()
+
+        PRINT()  # start on a fresh line
+        PRINT("=" * 80)
+        PRINT(f"APPLICATION_PROJECT_HOME == {cls.APPLICATION_PROJECT_HOME!r}")
+        PRINT(f"PYPROJECT_TOML_FILE == {cls.PYPROJECT_TOML_FILE!r}")
+        PRINT(f"PYPROJECT_NAME == {cls.PYPROJECT_NAME!r}")
+        the_app_project: Project = Project.app_project
+        the_app_project_class: Type[Project] = the_app_project.__class__
+        the_app_project_class_name: str = the_app_project_class.__name__
+        assert (Project.app_project
+                == app_project()
+                == the_app_project_class.app_project
+                == the_app_project.app_project), (
+            "Project consistency check failed."
+        )
+        PRINT(f"{the_app_project_class_name}.app_project == Project.app_project == app_project() == {app_project()!r}")
+        the_app: Project = app_project()
+        for attr in sorted(dir(the_app)):
+            val = getattr(the_app, attr)
+            if attr.isupper() and not attr.startswith("_") and (val is None or isinstance(val, str)):
+                PRINT(f"app_project().{attr} == {val!r}")
+        PRINT("=" * 80)
+
+    @classproperty
+    def app_project(cls) -> Project:  # noQA - PyCharm thinks we should use 'self'
+        """
+        Once the project is initialized, ProjectRegistry.app_project returns the application object
+        that should be used to dispatch project-dependent behavior.
+        """
+        app_project: Project = cls._shared_app_project_cell.value or cls.initialize()
+        return app_project
+
+
+Project.declare_project_registry_class(ProjectRegistry)  # Finalize a circular dependency
+
+
+# --------------------------------------------------------------------------------
+# C4-specific tools follow
+# --------------------------------------------------------------------------------
+
+
+class C4ProjectIdentity(ProjectIdentity):
+
+    @classmethod
+    def appify(cls, name: str) -> str:
+        return name.replace('-portal', '').replace('encoded-', '')
+
+    @classmethod
+    def repofy(cls, name: str) -> str:
+        return name.replace('dcic', '')
+
+    @classmethod
+    def prettify(cls, name: str) -> str:
+        return super().prettify(name).replace("Cgap", "CGAP").replace("Smaht", "SMaHT")
+
+
+class C4Project(Project):
+    """
+    Uses C4 functionality for naming. Outside organizations may not want such heuristics.
+    """
+
+    IDENTITY_CLASS = C4ProjectIdentity
+
+
+class C4ProjectRegistry(ProjectRegistry):
+
+    PROJECT_BASE_CLASS = C4Project
+
+    BAD_PYPROJECT_NAMES = [('cgap-portal', 'encoded'),
+                           ('fourfront', 'encoded'),
+                           ('smaht', 'encoded')]
```

Comment:
This isn't really the "bad" project names is it? It's the ones you have to use? I think the naming could be better.

Reply:
I renamed it to DISALLOWED.